[wip,dnm] scripts: fix gceworker.sh sync

irfansharif opened the pull request:
Fixing some crap. Assumes you have the right unison version locally and
stuff (ping @irfansharif if you want the right binary).

Forces sync one way (local -> gceworker), doesn't sync .git. Easiest to
start if you have a checkout remotely, and have run `git submodule
update --force --recursive --init` at least once.

Release note: None

adityamaru committed:
[wip,dnm] scripts: fix gceworker.sh sync

Fixing some crap. Assumes you have the right unison version locally and
stuff (ping @irfansharif if you want the right binary).

Forces sync one way (local -> gceworker), doesn't sync .git. Easiest to
start if you have a checkout remotely, and have run `git submodule
update --force --recursive --init` at least once.

Release note: None

Co-authored-by: irfan sharif <irfanmahmoudsharif@gmail.com>
Co-authored-by: Aditya Maru <adityamaru@gmail.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54590)
<!-- Reviewable:end -->

irfansharif commented:
You can also skip the submodule init I think, but it just takes a lot longer for unison to pick up all the submodule stuff from your local machine than it is from just the internet (especially all the yarn stuff).

irfansharif commented:
The .git directory on remote is pure junk though. You could be on a SHA locally, and all the files + submodules would appear correctly remotely corresponding to that SHA, but if you ever tried doing `git status` remotely, it’ll show you a diff from what it _thinks_ the SHA is (pure junk) to the SHA you have checked out locally.

irfansharif committed:
[wip,dnm] scripts: introduce gceworker.sh run '...'

To be used in conjunction with `sync`, runs arbitrary commands from
remote checkout of local workspace (including all unstaged diffs).

Release note: None

irfansharif commented:
I think I was able to get the sync thing working. Here’s a local diff that auto-propagates remotely where I was just running the test remotely. Sync is just happening in the background.

![image](https://user-images.githubusercontent.com/10536690/93650561-d632db80-f9dc-11ea-968c-60a77b819b78.png)

irfansharif commented:
```
brew uninstall unison
curl -L 'https://github.com/irfansharif/utils/raw/master/unison/unison-2.51.2.bottle.tar.gz' -o /tmp/unison-2.51.2.bottle.tar.gz
brew install -f /tmp/unison-2.51.2.bottle.tar.gz
brew pin unison
```

We can use the above for the right unison build going forward ("unison version 2.51.2 (ocaml 4.07.0)").

irfansharif commented:
Hm, one place where it's actually useful to have bi-directional sync is for auto-generated code. `make protobuf` is pretty fast usually, but it's a bit weird to have to run a subset of build commands remotely and another subset locally (like `make generate`).

irfansharif commented:
I'm not going to get to this any time soon. I still think it's a good idea in general, but there are too many UX hurdles here. Maybe the move to make here is use bazel to run remote tests (through #56965).
